src/storage/txn/commands/prewrite.rs

```diff
@@ -537,12 +537,11 @@ impl<K: PrewriteKind> Prewriter<K> {
                     txn.guards = Vec::new();
                     final_min_commit_ts = TimeStamp::zero();
                 }
-                e @ Err(MvccError(box MvccErrorInner::KeyIsLocked { .. })) => {
-                    locks.push(
-                        e.map(|_| ())
-                            .map_err(Error::from)
-                            .map_err(StorageError::from),
-                    );
+                Err(MvccError(box MvccErrorInner::KeyIsLocked { .. })) => {
+                    match check_committed_record_on_err(prewrite_result, txn, reader, &key) {
+                        Ok(res) => return Ok(res),
+                        Err(e) => locks.push(Err(e.into())),
+                    }
                 }
                 Err(e @ MvccError(box MvccErrorInner::AssertionFailed { .. })) => {
```

Comment:
@MyonKeminta @sticnarf 
Is it necessary to also cover the assertion check failures here for the possible retrying prewrite requests?

Reply:
This also reminds me about the last branch, other errors. It can be anything strange, for example, disk IO errors. Furthermore, it can be still not idempotent and it may be even impossible to do `check_committed_record_on_err`. In this case, it may result in a false commit failure...

Maybe we only report failure to the user when the error is of known types. For unknown errors, we treat them as undetermined errors.

Reply:
Agree, another thing that may need to be considered is how much overhead or impact would be introduced if we do more checks.

Reply:
The only case where the assertion failure may be reported before any conflict is found is as stated in this issue: https://github.com/tikv/tikv/issues/12113 . It seems adding check here may avoid throwing assertion_failed error in cases where the current transaction is in committed state. However it looks not necessary to me, since we already have a fix to #12113 .

Reply:
Will the idea about treating unknown errors as undetermined be done in this PR? By the way, do we have any injection for producing IO errors?

Reply:
> Will the idea about treating unknown errors as undetermined be done in this PR? By the way, do we have any injection for producing IO errors?

I can't remember we inject any IO errors in regular tests (otherwise, I believe we should have been discovered something strange before).

Treating unknown errors as undetermined will not be included in _this_ PR. It should be done in the client part, I think. And that problem is not a new issue, it seems to exist from the very beginning. So, that problem may be not as severe and urgent as this one.